### PR TITLE
tdnf: Add check to identify duplicate repo id

### PR DESCRIPTION
**Pull request by shivania2**
When a repository(Repo id) is set multiple times, tdnf should throw an error to prevent issues.

Recently a customer was disabling all photon* repos in the files in /etc/yum.repos.d, and replaced them with another repo config, which used the same repo ids ('photon', 'photon-updates', etc.). The URLs in the original repo configs  (pointing to packages.vmware.com) were not reachable. The URLs in the replaced repo configs were reachable. However, tdnf used the URLs from the original repo configs, which lead to timeouts. Disabling the repositories is not enough - tdnf still uses the settings, and the setting that is used is arbitrary. This issue could have been prevented if tdnf threw an error.

### CODE REVIEW

File: client/repoutils.c [hunk 1]
```diff
@@ -541,7 +541,9 @@ TDNFFindRepoById(
     )
 {
     uint32_t dwError = 0;
+    PTDNF_REPO_DATA parse_pRepos = NULL;
```
Comment (oliverkurth) [inline client/repoutils.c]: Nit: please adhere to naming standards - name it `pRepoParse` or something. Use a suffix, not a prefix.

File: client/repoutils.c [hunk 2]
```diff
@@ -553,18 +555,23 @@ TDNFFindRepoById(
...
+            nFound = 1;
+            pRepos = parse_pRepos;
+        } else if(!strcmp(pszRepo, parse_pRepos->pszId) && nFound == 1)
+        {
+            dwError = ERROR_TDNF_DUPLICATE_REPO_ID;
+            BAIL_ON_TDNF_ERROR(dwError);
         }
-        pRepos = pRepos->pNext;
+        parse_pRepos = parse_pRepos->pNext;
     }
```
Comment (oliverkurth) [inline client/repoutils.c]: This is the wrong place for this check, for two reasons: this function will be called too late, or may actually not be called at all. It's utility function to find a repo, it's not intended to check the configuration. So it may also be called multiple times, and since this adds extra CPU cycles, makes it inefficient.

A better place for the check would be in `TDNFLoadRepoData()` in `client/repolist.c` at the end. You can loop through all repos there and check for dups.

Besides all that, this logic also has a flaw - we'd only detect duplicates for repositories we are looking for. If for example we are looking for `photon-bar`, and `photon-foo` is a dup, we wouldn't see that.

File: pytests/tests/test_baseurls.py [hunk 3]
```diff
@@ -32,11 +32,37 @@ def teardown_test(utils):
...
+    filename = os.path.join(utils.config['repo_path'], "yum.repos.d", REPOFILENAME)
+
+    baseurls = "http://localhost:8080/photon-test"
+    utils.create_repoconf(filename, baseurls, reponame)
+
+    ret = utils.run(['tdnf',
+                     '--disablerepo=*', '--enablerepo={}'.format(reponame),
+                     'makecache'],
+                    cwd=workdir)
+    assert ret['retval'] == 0
```
Comment (oliverkurth) [inline pytests/tests/test_baseurls.py]: If we have multipe repos with the same id, `makecache` should already throw an error, so you should assert for `1037` here.

File: pytests/tests/test_baseurls.py [hunk 4]
```diff
@@ -32,11 +32,37 @@ def teardown_test(utils):
...
-def test_multiple_baseurls(utils):
+def test_multiple_repoid(utils):
     reponame = TESTREPO
     workdir = WORKDIR
     utils.makedirs(workdir)
 
+    filename = os.path.join(utils.config['repo_path'], "yum.repos.d", REPOFILENAME)
+
+    baseurls = "http://localhost:8080/photon-test"
+    utils.create_repoconf(filename, baseurls, reponame)
```
Comment (oliverkurth) [inline pytests/tests/test_baseurls.py]: Does this create another repo with an id that already exists? That's not clear to me.

File: pytests/tests/test_baseurls.py [hunk 5]
```diff
@@ -32,11 +32,37 @@ def teardown_test(utils):
         os.remove(filename)
 
 
-def test_multiple_baseurls(utils):
+def test_multiple_repoid(utils):
```
Comment (oliverkurth) [inline pytests/tests/test_baseurls.py]: Why is this test in `test_baseurls.py`? It has nothing to do with multiple base URLs.

**Review by oliverkurth — CHANGES_REQUESTED** [review]

File: client/repoutils.c (hunk 2 above)
Comment (shivania2) [reply client/repoutils.c]: Added check to identify duplicate repo id in TDNFLoadRepoData() in client/repolist.c

File: pytests/tests/test_baseurls.py (hunk 5 above)
Comment (shivania2) [reply pytests/tests/test_baseurls.py]: Added test case in test_repolist.py

File: pytests/tests/test_baseurls.py (hunk 3 above)
Comment (shivania2) [reply pytests/tests/test_baseurls.py]: Done

File: pytests/tests/test_baseurls.py (hunk 4 above)
Comment (shivania2) [reply pytests/tests/test_baseurls.py]: Yes, This creates another repo with the id [cfg-test], which already exist. I have removed this code in latest changes

File: client/repoutils.c (hunk 1 above)
Comment (shivania2) [reply client/repoutils.c]: Done

File: client/repolist.c [hunk 6]
```diff
@@ -115,18 +127,35 @@ TDNFLoadRepoData(
...
+    dwError = TDNFAllocateMemory(nRepoId + 1, sizeof(char *), (void **)&ppRepoId);
+    BAIL_ON_TDNF_ERROR(dwError);
+
+    pRepoParse = pReposAll;
+
+    while (pRepoParse != NULL) {
+        uint32_t hash = hashString(pRepoParse->pszId) % nRepoId;
+        if (ppRepoId[hash] != NULL && strcmp(ppRepoId[hash], pRepoParse->pszId) == 0) {
+            dwError = ERROR_TDNF_DUPLICATE_REPO_ID;
+            BAIL_ON_TDNF_ERROR(dwError);
```
Comment (oliverkurth) [inline client/repolist.c]: We don't need to use hashes. Just loop twice, it's okay, even when it's O(n^2) - there aren't that many repo configs.

File: pytests/tests/test_baseurls.py [hunk 7]
```diff
@@ -33,7 +33,7 @@ def teardown_test(utils):
 
 
 def test_multiple_baseurls(utils):
-    reponame = TESTREPO
```
Comment (oliverkurth) [inline pytests/tests/test_baseurls.py:36]: Why do you change this?

File: pytests/tests/test_repolist.py [hunk 8]
```diff
@@ -126,3 +127,23 @@ def test_repolist_invalid(utils):
 def test_repolist_memcheck(utils):
     ret = utils.run_memcheck(['tdnf', 'repolist'])
     assert ret['retval'] == 0
+
+
+# multiple repoid
+def test_multiple_repoid(utils):
+    reponame = 'bar1.repo'
+    repofile_bar1 = os.path.join(utils.config['repo_path'], 'yum.repos.d', reponame)
+    utils.edit_config(
```
Comment (oliverkurth) [inline pytests/tests/test_repolist.py:137]: Wouldn't you need to add the repo with id `bar` twice to get the error?

**Review by oliverkurth — CHANGES_REQUESTED** [review]

File: client/repolist.c [hunk 9]
```diff
@@ -115,12 +117,27 @@ TDNFLoadRepoData(
 
         TDNF_SAFE_FREE_MEMORY(pszRepoFilePath);
         pszRepoFilePath = NULL;
-
         /* may have added multiple repos, go to last one */
         while (*ppRepoNext)
             ppRepoNext = &((*ppRepoNext)->pNext);
     }
 
+    pRepoParsePre = pReposAll;
```
Comment (sshedi) [inline client/repolist.c]:
```
for (pRepoParsePre = pReposAll;
        pRepoParsePre;
        pRepoParsePre = pRepoParsePre->pNext) {
    for (pRepoParseNext = pRepoParsePre->pNext;
            pRepoParseNext;
            pRepoParseNext = pRepoParseNext->pNext) {
        if (!strcmp(pRepoParsePre->pszId, pRepoParseNext->pszId)) {
            dwError = ERROR_TDNF_DUPLICATE_REPO_ID;
            BAIL_ON_TDNF_ERROR(dwError);
        }
    }
}
```

I would suggest checking of baseurl instead of ID. ID is not a primary key but baseurl is.
If baseurls are same, throw duplicate error.

**Review by sshedi — CHANGES_REQUESTED** [review]

File: client/repolist.c (hunk 9 above)
Comment (oliverkurth) [reply client/repolist.c]: No, that's just plain wrong. The id is the primary key. There is no issue at all if a `baseurl` is duplicated - it's not advisable, but not really an issue.

Having duplicate ids however is a problem.That is why we are doing this.

File: pytests/tests/test_baseurls.py (hunk 7 above)
Comment (oliverkurth) [reply pytests/tests/test_baseurls.py:36]: Can you please revert this? I think you left it over accidentally.

File: client/repolist.c [hunk 10]
```diff
@@ -115,12 +117,27 @@ TDNFLoadRepoData(
...
+        while (pRepoParseNext->pNext != NULL) {
+            if (!strcmp(pRepoParsePre->pszId, pRepoParseNext->pNext->pszId)) {
+                dwError = ERROR_TDNF_DUPLICATE_REPO_ID;
+                BAIL_ON_TDNF_ERROR(dwError);
+            }
+            else
+                pRepoParseNext = pRepoParseNext->pNext;
+        }
+        pRepoParsePre = pRepoParsePre->pNext;
+    }
```
Comment (oliverkurth) [inline client/repolist.c]: Thank you, this looks good. I liked the version using `for` better though because it makes the algorithm easier to understand (and I can only see it from Shreenidhi's comment).

File: client/repolist.c [hunk 11]
```diff
@@ -115,12 +117,27 @@ TDNFLoadRepoData(
...
 
+    pRepoParsePre = pReposAll;
+
+    while (pRepoParsePre != NULL && pRepoParsePre->pNext != NULL) {
+        pRepoParseNext = pRepoParsePre;
+ 
+        while (pRepoParseNext->pNext != NULL) {
+            if (!strcmp(pRepoParsePre->pszId, pRepoParseNext->pNext->pszId)) {
+                dwError = ERROR_TDNF_DUPLICATE_REPO_ID;
+                BAIL_ON_TDNF_ERROR(dwError);
```
Comment (oliverkurth) [inline client/repolist.c:131]: Please also add a `pr_err` with an error message that also prints the id, so we know exactly which id is duplicated.

**Review by oliverkurth — CHANGES_REQUESTED** [review]
This looks good, just minor comments.

I think the error in the tests is because of the change you left over in `test_baseulrd.py`, even though the failure is in `test_setopt_reposdiir.py` - this may be in bug not from this PR. So please revert that and try again.

File: client/repolist.c [hunk 12]
```diff
@@ -115,12 +117,22 @@ TDNFLoadRepoData(
...
         /* may have added multiple repos, go to last one */
         while (*ppRepoNext)
             ppRepoNext = &((*ppRepoNext)->pNext);
     }
 
+    for (pRepoParsePre = pReposAll; pRepoParsePre; pRepoParsePre = pRepoParsePre->pNext) {
+
+        for (pRepoParseNext = pRepoParsePre->pNext; pRepoParseNext; pRepoParseNext = pRepoParseNext->pNext) {
+            if (!strcmp(pRepoParsePre->pszId, pRepoParseNext->pszId)) {
+                pr_err("ERROR: Duplicate Repoid (%s)\n", pRepoParsePre->pszId);
```
Comment (oliverkurth) [inline client/repolist.c]: Nit: better is "duplicate repo id: '%s'".

**Review by oliverkurth — COMMENTED** [review]
Thanks again! Just a tiny nit.

We need to check why the tests are failing - it looks to me that we already had duplicate repo ids in the test and we are only noticing them now.

**Comment by oliverkurth** [comment]
> We need to check why the tests are failing - it looks to me that we already had duplicate repo ids in the test and we are only noticing them now.

That was indeed the issue. I have a fix here: https://github.com/oliverkurth/tdnf/commit/a859f3399bc4aec7ba5f042f40fd9ab0f56bdbd8 . You can apply those changes to yours.

Proof that that succeeds: https://github.com/oliverkurth/tdnf/actions/runs/6356463459

**Review by oliverkurth — APPROVED** [review]
